Add custom exceptions to libdnf interface

lukash opened the pull request:
This one certainly turned out bigger than I anticipated, and still is not a complete solution in some aspects (mainly on the C API, where eror handling is limited by the current interface).

This PR adds exception classes for common, expected errors and also for unexpected errors (although those don't really matter much and whatever exceptions are used for those, I've left them in place). For the expected errors (e.g. file not found, corrupted SQLite database), it is important to have an exception base that is distinguishable from the others. The PR adds `libdnf::Error` as such base exception class and then adds handling of the errors on all the various interfaces (C, Hawkey, SWIG).

Note this currently raises the requirement for cmake to version 3.8, as noted in a TODO in the relevant commit, so we most likely don't want to merge this as-is.

Conan-Kudo commented:
@lukash I'm actually of the opinion that we should just bite the bullet and raise the CMake version. For EL7 builds, we can just depend on cmake3 from EPEL, and that should be sufficient. Nowadays, we care primarily about Fedora, EL8, Mageia, and openSUSE. None of those don't already ship CMake > 3.8.

Conan-Kudo commented:
@lukash Could you please rebase this?

lukash commented:
> @lukash I'm actually of the opinion that we should just bite the bullet and raise the CMake version. 

@Conan-Kudo I agree with that myself. I can't imagine why we would want to package this for RHEL7. Also, I'd have to copy-paste a ~50 line CMake macro to make this work with old CMake (though I haven't tested it yet, so it may get more complicated). That can also be done when (if) we need to build for RHEL7. But I'll need to discuss this with the team when they're back from the holidays.

> For EL7 builds, we can just depend on cmake3 from EPEL, and that should be sufficient.

Not sure that can be done, as @dmach told me we need CMake 2.8 for RHEL7... But I'll bring that up when we discuss it, thanks.

dmach commented:
@lukash wouldn't bundling /usr/share/cmake/Modules/UseSWIG.cmake in libdnf/cmake/modules-cmake-2/ do the job as well? Can you try if it compiles on RHEL 7?
(DNF 4 stack is supported on RHEL 7 and we ideally shouldn't break the compatibility)

lukash commented:
@dmach well.. I was thinking if I copy it from the 3.8+ version, it may not to work with CMake 2.8 on RHEL7. And if I copy the version from CMake 2.8, it again may not work with CMake 3.8 and also I'll need to alter it to support that `SHARED` type.

But I've just found out there already is `libdnf/cmake/modules-cmake-2/UseSWIG.cmake` in the repo. And it seems it's only used for CMake 2, which should solve the version issues, so I may just need to alter it to support the `SHARED` type (and make it the same syntax as what is used for CMake 3). While I don't particularly like altering copy-pasted third-party code, it should solve the problem.

lukash commented:
I've rebased to fix the conflicts with the version changes in master.

I've also tried to build libdnf on RHEL 7 to get the CMake changes working and verify them, but there are missing dependencies on RHEL 7 which are hard to get there. In the end we decided to bump the CMake version and not attempt to keep this RHEL 7 compatible atm. (but we will not make further changes/improvements without a good reason, to keep it reasonably compatible for the case we need to backport this to RHEL 7 later).

So I've removed the TODO, bumped the CMake version and am using the new `swig_add_library` function (which is what would have to be done differently for RHEL7).

Conan-Kudo commented:
@lukash Do you think there's a way to refactor the SWIG stuff so that it'd be easy to add new bindings with minimal effort? For example, adding Ruby bindings should be technically easy, since all the SWIG features we need are present (including `shared_ptr`). But this current approach seems to make it harder to add new bindings...

lukash commented:
@Conan-Kudo tbh. I have no idea. I am new to SWIG and it's quite complex. The discussion should probably happen elsewhere... I'm not sure where, maybe create an issue for it unless you think of a better place? If you do, please mention the current issues.

I'll just mention here that's most probably something we'll consider for dnf-5 and it will depend on the amount of specific code it needs to support that particular language. And f we add the bindings it means we're somewhat committing to support them too... I mean if it just worked it would be great, but in case of issues we'd need a way to test it etc...?

jrohel commented:
@rh-atomic-bot r+

lukash commented:
@rh-atomic-bot retry

lukash commented:
There was a problem with the python modules not actually finding the `_error.so` module they were linked against - in my test setup it worked, but normally the dynamically linked libraries aren't being looked up in the python modules directories, where the `_error.so` is installed.

What fixed this is importing `libdnf.error` first in `libdnf/__init__.py`, which causes the `_error.so` to be loaded when the global variable (the python exception pointer) symbol is needed and it can be used. For some reason, though, the modules still need to be linked against `_error.so` as well, otherwise it errors out with `undefined symbol: m_Error`.

I've looked into the `_common_types.so` (one of the users of the exception) symbol table and the `m_Error` is undefined in both (linked and non-linked) variants, with a slight difference of having type `OBJECT` in the linked module and `NOTYPE` in the non-linked, which shouldn't matter much from what I understand. I suspect it's something about the module actually being linked against `_error.so` that makes it find the symbol when `_error.so` is actually loaded (otherwise it would proceed to load it and fail with not finding the file).

The whole mechanism feels quite incidental at this point and I'm reluctant to use it in its current state, but at the same time I have no better ideas (besides putting the bindings into a single module or putting the python exception pointer somehow into libdnf.so itself... neither of which is particularly compelling).

lukash commented:
I've changed the way the global variable symbol was loaded to calling `sys.setdlopenflags(os.RTLD_NOW | os.RTLD_GLOBAL)` in python before the `libdnf.error` module is imported (and resetting the dlopen flags back to the default right after loading the module).

It seems a bit cleaner and I was able to remove the linking of the other modules against `_error.so`, as well as the CMake version bump, which was required just because of this.

I've also renamed the variable (and therefore the symbol) to `libdnf_error`.

jrohel commented:
@rh-atomic-bot retry

jrohel commented:
@rh-atomic-bot r+

jrohel commented:
@lukash Fix libdnf version. The 0.44.0 was used by your another PR.

lukash commented:
Yes... annoying.

Also one scenario is failing on some exceptions, seems I've broken that, I need to look into it.

lukash commented:
The first force-push is a rebase, the second one is a fix for the failing test, which exposed a problem with registering the exception handlers in SWIG - the handlers weren't registered due to some weirdness in how SWIG handles its `%includes` and `%imports`. I've added some comments into the code too, I'm afraid this won't be the first time we're running into this.

jrohel commented:
@rh-atomic-bot r+
